### ISSUE TITLE
Large WinRM uploads failing

### ISSUE THREAD

**Issue by dylanmei** [packer-community/packer-windows-plugins#3]
Large file uploads, such as the vmware tools, end abruptly.

    Build 'vmware-windows-iso' errored: unexpected EOF

In my case, the temporary file at C:\Users\packer\AppData\Local\Temp repeatedly gets to 15.4 MB (16,211,663 bytes) when it quits.

**Comment by dylanmei** [comment]
The problems specific to this issue are addressed in 199e29fb307659a3792bcfcaa1966bf16a7d7734

All file-copying stuff is moved into github.com/packer-community/winrmcp. I've made sure it

1. uses less memory and
1. keeps using a winrm.Shell until it reaches the `MaxConcurrentOperationsPerUser` limit, at which point it creates the new winrm.Shell.

The second point makes it mildly quicker than before. A 37mb file went from 3:30 minutes to 2:30 minutes, and a 90mb file from 8:00 to 6:00.

I've only tinkered at it, but running uploads simultaneously makes the directory uploads quite fast. If we can get that stable then we should be able to break up the larger files into parallel uploads as well.

**Comment by sneal** [comment]
@dylanmei Since MaxConcurrentOperationsPerUser is configurable and the default varies between OS versions, we're planning on catching the associated error and retrying after resetting the connection in the Ruby WinRM gem.

**Comment by dylanmei** [comment]
@sneal Cool!

Winrmcp reads the setting ahead of time: https://github.com/packer-community/winrmcp/blob/master/winrmcp/cp.go#L56

I can see utilizing other settings to pro-actively warn about MaxMemoryPerShell problems and to enable running multiple shells in parallel.

**Comment by sneal** [comment]
@mwrock What do you think about reading the WinRM MaxConcurrentOperationsPerUser like @dylanmei is doing here in the Ruby gem? That seems a bit more robust.

**Comment by mwrock** [comment]
Interesting. That would be ideal. I had tried doing that in the Test-Kitchen implementation and hit a wall with access denied errors. I was unable to read the winrm info on the guest side authenticating remotely with an administrator account. I also tried using plain windows powershell remoting and concluded this was likely a blocked call for remote tokens.

@dylanmei are you using a scheduled task to call via a local context? That would work but was more than I wanted to bite off at the moment. And if you are not, then there is something I was obviously missing.

**Comment by mwrock** [comment]
Just looked at the code and there is clearly no scheduled task foo going on so I will look at this again. Way better than using exception handling for control logic.

**Comment by dylanmei** [comment]
@mwrock Well this is curious. I've run [winrmcp](https://github.com/packer-community/winrmcp) with this technique only on Win 2012 R2 via Packer vmware and aws. I have fairly familiar WinRM settings [(gist)](https://gist.github.com/dylanmei/cead520633e629db321e).

I may be being too cavalier. If you do find any kind of repro, please do report back here so I can create an issue for winrmcp. I'll do the same if I spot something.

**Comment by mefellows** [comment]
As an aside, an SMB share is probably worth investigating as an alternative as most hosts and guests support CIFS/SMB shares by default and is the most native Windows sharing mechanism that I can think of. As it turns out BITS looks more trouble that it's worth (involves enabling IIS and the BITS extension for uploads to work which raises some security concerns not to mention complicates setup and cleanup activities).

**Comment by mwrock** [comment]
Yeah a share could serve as an efficient and convenient transport mechanism. Unfortunately it would not be feasible universally since it assumes the host is addresable by the guest which is not likely to be the case for cloud based scenarios. I think you eluded to that above.

**Comment by mefellows** [comment]
Thanks @mwrock, could we not create the share on the remote host first and then use standard tools to authenticate and copy the files to the remote share?

**Comment by mwrock** [comment]
@dylanmei and @sneal: I was just investigating the remote invocation of `winrm get winrm/config -format:xml`. This does indeed work on my 2012R2/win8.1 VMs. But it fails with:

```
<f:WSManFault xmlns:f="http://schemas.microsoft.com/wbem/wsman/1/ws
manfault" Code="5" Machine="wrocktest2k8r2"><f:Message>Access is denied. </f:Me
ssage></f:WSManFault>
```

On win7 and 2k8R2 VMs which, sadly, is were one most needs this info from with its default limit of 15. I've tried several permutations of this command including the navigation of the `wsman:\ ` powershell provider with no luck. I've done a fair amount of googling to get a more solid answer but everything addresses folks trying to enable remoting and not just checking remoting configuration. I also tried temporarily turning off the windows firewall ithout any success.

At any rate, my guess is that the security model changed between powershell 2 and 3 preventing winrm config commands from being remotely invoked on powershell v2 and earlier.
